rpmlint/ReadelfParser.py

```diff
@@ -0,0 +1,126 @@
+  [11] .shstrtab         STRTAB          0000000000000000 000208 000059 00      0   0  1
+Key to Flags:
+  W (write), A (alloc), X (execute), M (merge), S (strings), I (info),
+  L (link order), O (extra OS processing required), G (group), T (TLS),
+  C (compressed), x (unknown), o (OS specific), E (exclude),
+  l (large), p (processor specific)
+"""
+
+class ElfSectionInfo:
+    """
+    Class contains information about ELF sections of an ELF file. The information
+    is get with the following command line: readelf -WS.
+    """
+
+    section_regex = re.compile(r'.*\] (?P<section>[^\s]*)\s*\w+\s*\w*\s*\w*\w*\s*(?P<size>\w*)')
+
+    def __init__(self, path, issues):
+        self.path = path
+        self.issues = issues
+        self.is_archive = self.path.endswith('.a')
+        self.files = []
+        self.parse()
+
+        # register all check functions
+        self.check_functions = [self.check_lto_section, self.check_no_text_in_archive]
+
+    def parse(self):
+        r = subprocess.check_output(['readelf', '-W', '-S', self.path], encoding = 'utf8')
+        lines = [l for l in r.split('\n')]
+
+        needle = 'Section Headers:'
+
+        # archive files can contain multiple files
+        while len(lines) > 0:
+            parsed_sections = []
+            lines = list(dropwhile(lambda x: not needle in x, lines))
+
+            # skip header and empty section
+            lines = lines[3:]
+
+            sections = list(takewhile(lambda x: not 'Key to Flags:' in x, lines))
+            for s in sections:
+                r = ElfSectionInfo.section_regex.search(s)
+                assert r != None
+                parsed_sections.append(ElfSection(r.group('section'), r.group('size')))
+
+            lines = lines[len(sections):]
+            if len(parsed_sections) > 0:
+                self.files.append(parsed_sections)
+
+    # For an archive, test if any .text sections is empty
+    def check_no_text_in_archive(self):
+        if self.is_archive:
+            for f in self.files:
+                for section in f:
+                    if section.name == '.text' and section.size == 0:
+                        self.issues.append(Issue('E', 'lto-no-text-in-archive', self.path))
+                        return
+
+    # Check for LTO sections
+    def check_lto_section(self):
+        for f in self.files:
+            for section in f:
+                if '.gnu.lto_.' in section.name:
+                    self.issues.append(Issue('E', 'lto-bytecode', self.path))
+                    return
+
+    def check(self):
+        for fn in self.check_functions:
+            fn()
+
+class ReadelfParser:
+    """
+    Class contains all information obtained by readelf command
+    in a structured format.
+    """
+
+    def __init__(self, path):
+        self.issues = []
+        self.section_info = ElfSectionInfo(path, self.issues)
+
+    def check(self):
```

Comment:
Does this need to be 2 step init+function, or when we start readelfparser we simply want it to process everything right away?